src/cmd/promremotebench/query.go

```diff
@@ -237,24 +245,83 @@ func (q *queryExecutor) alertLoad(checker Checker) {
 				mustWriteString(query, "})")
 			}
 
-			q.executeQuery(query, false, q.LoadRange, q.LoadStep)
+			q.fanoutQuery(query, false, q.LoadRange, q.LoadStep)
 		}()
 	}
 }
 
-func (q *queryExecutor) executeQuery(query *strings.Builder, retResult bool, queryRange, queryStep time.Duration) []byte {
+func (q *queryExecutor) fanoutQuery(
+	query *strings.Builder,
+	retResult bool,
+	queryRange time.Duration,
+	queryStep time.Duration,
+) ([][]byte, error) {
 	now := time.Now()
 	values := make(url.Values)
 	values.Set("query", query.String())
 	values.Set("start", strconv.Itoa(int(now.Add(-1*queryRange).Unix())))
 	values.Set("end", strconv.Itoa(int(now.Unix())))
 	values.Set("step", queryStep.String())
 
-	reqURL := fmt.Sprintf("%s?%s", q.URL, values.Encode())
+	var (
+		wg       sync.WaitGroup
+		mu       sync.Mutex
+		multiErr xerrors.MultiError
+		qs       = values.Encode()
+	)
+
+	results := make([][]byte, 0, len(q.URLs))
+	for _, url := range q.URLs {
+		wg.Add(1)
+		reqURL := fmt.Sprintf("%s?%s", url, qs)
+
+		if q.Debug {
+			q.Logger.Info("fanout query",
+				zap.String("url", reqURL),
+				zap.Any("values", values))
+		}
+
+		go func() {
+			res, err := q.executeQuery(reqURL, retResult)
+			mu.Lock()
+			multiErr = multiErr.Add(err)
+			results = append(results, res)
+			mu.Unlock()
+		}()
+	}
+
+	wg.Wait()
+
+	if err := multiErr.FinalError(); err != nil {
+		q.Logger.Error("fanout error", zap.Error(err))
+		return results, err
+	}
+
+	// NB: If less than 2 results returned, no need to compare for equality.
+	if !retResult || len(results) < 2 {
+		return results, nil
+	}
+
+	firstResult := results[0]
+	for i, res := range results[1:] {
+		if bytes.Equal(res, firstResult) {
+			continue
+		}
+
+		q.Logger.Error("mismatch in returned data", zap.Int("index", i))
```

Comment:
Should we log the mismatching data here?

Reply:
Not initially, I think; they're likely to be huge byte buffers which won't really be readable anyway. If we start seeing this failing a lot, maybe we could parse them out into actual query results for debugging, but this is probably fine for now